cpp/include/raft/core/device_resources_snmg_nccl.hpp

```diff
@@ -0,0 +1,201 @@
+    : device_resources_snmg(clique)
+  {
+  }
+
+  device_resources_snmg_nccl(device_resources_snmg_nccl&&)            = delete;
+  device_resources_snmg_nccl& operator=(device_resources_snmg_nccl&&) = delete;
+
+  ~device_resources_snmg_nccl()
+  {
+    std::vector<raft::resources>& clique = raft::resource::get_multi_gpu_resource(*this);
+    int num_ranks                        = clique.size();
+
+    ncclGroupStart();
+    for (int rank = 0; rank < num_ranks; rank++) {
+      ncclComm_t& nccl_comm = raft::resource::get_nccl_comm(clique[rank]);
+      RAFT_NCCL_TRY_NO_THROW(ncclCommDestroy(nccl_comm));
+    }
+    ncclGroupEnd();
+  }
+
+  /**
+   * @brief Set root rank of NCCL clique
+   */
+  inline void set_nccl_root_rank(int rank) { raft::resource::set_root_rank(*this, rank); }
+
+  /**
+   * @brief Get root rank of NCCL clique
+   */
+  inline int get_nccl_root_rank() const { return raft::resource::get_root_rank(*this); }
+
+  /**
+   * @brief Get number of ranks in NCCL clique
+   */
+  inline int get_nccl_num_ranks() const { return raft::resource::get_num_ranks(*this); }
+
+  /**
+   * @brief Get device ID of rank in NCCL clique
+   */
+  inline int get_device_id_of_rank(int rank) const
+  {
+    const raft::resources& dev_res = raft::resource::get_device_resources_for_rank(*this, rank);
+    return raft::resource::get_device_id(dev_res);
+  }
+
+  /**
+   * @brief Get NCCL comm object of rank in NCCL clique
+   */
+  inline ncclComm_t& get_nccl_comm_for_rank(int rank) const
+  {
+    const raft::resources& dev_res = raft::resource::get_device_resources_for_rank(*this, rank);
+    return raft::resource::get_nccl_comm(dev_res);
+  }
+
+  /**
+   * @brief Get raft::resources object of rank in NCCL clique
+   */
+  inline const raft::resources& get_device_resources_for_rank(int rank) const
+  {
+    return raft::resource::get_device_resources_for_rank(*this, rank);
+  }
+
+  /**
+   * @brief Set current device ID to rank and return its raft::resources object
+   */
+  inline const raft::resources& set_current_device_to_rank(int rank) const
+  {
+    return raft::resource::set_current_device_to_rank(*this, rank);
+  }
+
+  /**
+   * @brief Set current device ID to root rank and return its raft::resources object
+   */
+  inline const raft::resources& set_current_device_to_root_rank() const
+  {
+    return raft::resource::set_current_device_to_root_rank(*this);
+  }
+
+  /**
+   * @brief Set a memory pool on all GPUs of the clique
+   */
+  void set_memory_pool(int percent_of_free_memory) const
+  {
+    for (int rank = 0; rank < get_nccl_num_ranks(); rank++) {
+      const raft::resources& dev_res = set_current_device_to_rank(rank);
+      // check limit for each device
+      size_t limit = rmm::percent_of_free_device_memory(percent_of_free_memory);
+      raft::resource::set_workspace_to_pool_resource(dev_res, limit);
+    }
+    cudaSetDevice(this->main_gpu_id_);
+  }
```

Comment:
This function could be useful for accelerating multi-GPU algorithms, regardless of NCCL usage. I recommend adding it to the `device_resource_snmg` class.

Reply:
Agreed. This should exist in the base `device_resources_snmg`.